// awesome-react-org-chart/dist/core/LayoutStrategyBase.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var BranchParentAlignment_1 = __importDefault(require("./BranchParentAlignment"));
var ConnectorAlignment_1 = __importDefault(require("./ConnectorAlignment"));
var LayoutStrategyBase = /** @class */ (function () {
    function LayoutStrategyBase() {
        this.ConnectorAlignment = ConnectorAlignment_1.default.Edge;
        this.ConnectorThickness = 0;
        this.ParentAlignment = BranchParentAlignment_1.default.InvalidValue;
        this.ParentChildSpacing = 20;
        this.ParentConnectorShield = 50;
        this.SiblingSpacing = 20;
        this.ChildConnectorHookLength = 5;
    }
    Object.defineProperty(LayoutStrategyBase.prototype, "SupportsAssistants", {
        get: function () {
            return this.GetSupportsAssistants();
        },
        enumerable: false,
        configurable: true
    });
    return LayoutStrategyBase;
}());
exports.default = LayoutStrategyBase;
